boot/bootutil/src/loader.c

```diff
@@ -1174,20 +1174,108 @@ boot_validated_swap_type(struct boot_loader_state *state,
 }
 #endif
 
+/* Helper macro to avoid compile errors with systems that do not
+ * provide function to check device type.
+ * Note: it used to be inline, but somehow compiler would not
+ * optimize out branches that were impossible when this evaluated to
+ * just "true".
+ */
+#ifdef MCUBOOT_DEV_WITHOUT_ERASE_THEN_NO_ERASE
+#define device_requires_erase(fa) (flash_area_erase_required(fa))
+#else
+#define device_requires_erase(fa) (true)
+#endif
+
 /**
- * Erases a region of flash.
+ * Erases a region of device that requires erase prior to write; does
+ * nothing on devices without erase.
  *
- * @param flash_area           The flash_area containing the region to erase.
+ * @param fap                   The flash_area containing the region to erase.
  * @param off                   The offset within the flash area to start the
- *                                  erase.
+ *                              erase.
  * @param sz                    The number of bytes to erase.
  *
  * @return                      0 on success; nonzero on failure.
  */
 int
 boot_erase_region(const struct flash_area *fap, uint32_t off, uint32_t sz)
 {
-    return flash_area_erase(fap, off, sz);
+    if (device_requires_erase(fap)) {
+        return flash_area_erase(fap, off, sz);
+    }
+    return 0;
+}
+
+/**
+ * Removes data from specified region either by writing 0x00 in place of
+ * data or by doing erase, if device has such hardware requirement.
+ *
+ * @param fa                    The flash_area containing the region to erase.
+ * @param off                   The offset within the flash area to start the
+ *                              erase.
+ * @param size                  The number of bytes to erase.
+ *
+ * @return                      0 on success; nonzero on failure.
+ */
+int
+boot_scramble_region(const struct flash_area *fa, uint32_t off, uint32_t size)
+{
+    int ret = 0;
+
+    if (device_requires_erase(fa)) {
+        return flash_area_erase(fa, off, size);
+    } else {
+        uint8_t buf[BOOT_MAX_ALIGN];
+        size_t size_done = 0;
+
+        memset(buf, flash_area_erased_val(fa), sizeof(buf));
+
+        while (size_done < size) {
+            ret = flash_area_write(fa, size_done + off, buf, sizeof(buf));
+            if (ret != 0) {
+                break;
+            }
+            size_done += sizeof(buf);
+        }
```

Comment:
this just assumes max align will not overflow the size but makes no checks to ensure that, it should reduces buffer on final write to ensure it does not overflow

Reply:
Yes, that was intentional as I have assumes that MCUboot should write at BOOT_MAX_ALIGN pace anyway, but maybe that should be fixed.